cibuildwheel/oci_container.py

```diff
@@ -11,15 +11,44 @@
 import typing
 import uuid
 from collections.abc import Mapping, Sequence
+from dataclasses import dataclass
 from pathlib import Path, PurePath, PurePosixPath
 from types import TracebackType
 from typing import IO, Dict
 
 from ._compat.typing import Literal
 from .typing import PathOrStr, PopenBytes
-from .util import CIProvider, detect_ci_provider
+from .util import CIProvider, detect_ci_provider, parse_key_value_string
 
-ContainerEngine = Literal["docker", "podman"]
+ContainerEngineName = Literal["docker", "podman"]
+
+
+@dataclass(frozen=True)
+class OCIContainerEngineConfig:
+    name: ContainerEngineName
+    create_args: Sequence[str] = ()
+
+    @staticmethod
+    def from_config_string(config_string: str) -> OCIContainerEngineConfig:
+        config_dict = parse_key_value_string(config_string, ["name"])
+        name = " ".join(config_dict["name"])
+        if name not in ["docker", "podman"]:
+            msg = f"unknown container engine {name}"
+            raise ValueError(msg)
+
+        name = typing.cast(ContainerEngineName, name)
+        # some flexibility in the option name to cope with TOML conventions
+        create_args = config_dict.get("create_args") or config_dict.get("create-args") or []
```

Comment:
Do we do this anywhere else? I'm pretty sure we decided not to, and instead just have nice error messages if it fails. There's not much reason to add this when you don't have to worry about backward compatibility. It's better to have one correct way instead of multiple nearly identical ways. It's easier for a third party to parse, for example, if there's one and only one way to do it. For example, I am often code-searching to see what people are doing in their pyproject.toml. It makes it much harder to grep if there's not a clear way to spell it.

This is a bit different since it's combing from inside an environment variable or from pyproject.toml. I'm not as worried about it when it comes from an environment variable, but it would at least be nice to not have ambiguity when it comes from pyproject.toml.

Reply:
This wasn't intended to be a change in policy for options, it's more that because the way options parsing works, the option is transformed from a dict to a string in the parser, so it would be a bit more code and more fiddly to be strict about it, and I didn't think it was worth it :) . I'm documenting the kebab-case in TOML and snake_case in env vars, so I figured accidental usage would be rare.

Reply:
Okay, fair enough.